devito/function.py

```diff
@@ -1010,8 +1010,15 @@ def argument_defaults(self, alias=None):
 
         :param alias: (Optional) name under which to store values.
         """
-        args = super(SparseFunction, self).argument_defaults(alias=alias)
-        args.update(self.coordinates.argument_defaults())
+        key = alias or self
+
+        args = ArgumentMap({key.name: self._data_buffer})
+
+        # Collect default dimension arguments from all indices
+        for i, s, o, k in zip(self.indices, self.shape, self.staggered, key.indices):
```

Comment:
I'm again confused here: what happens if you just do:
```
args.update(k.argument_defaults(size=s+o)
```

*If* this worked, then I *think* we could move the fix to the superclass (as I don't understand why is this fix necessary here, but not in the superclass, ie for all tensor functions)

Reply:
Because unlike normal Functions that only have indices derived from a grid object that always names its dimensions the same, Sparse functions indices are specific to each one (`p_%s`%name`) so without dimension aliasing `key.indices`, the argument are not correct.

Reply:
I don't understand yet. I tried the proposed change above, and all of the new tests still pass.

Reply:
`TestArguments.test_override_sparse_data_default_dim` doesn't:

```    
        # Check all argument are present
        for p in self.parameters:
            if p.name not in arguments:
>               raise ValueError("No value found for parameter %s" % p.name)
E               ValueError: No value found for parameter p_src1_e
```
as without dimension name aliasing, all end up in `p_src_2_*`

Reply:
> Because unlike normal Functions that only have indices derived from a grid object that always names its dimensions the same

I disagree with this though. For example, when injecting a source, ``u`` gets indexed indirectly. Some other times, one may use subdimensions to index into a Function; users may also start using extra indices for whatever dumb reason. So, I'm not sure why should this fix stay in `SparseFunction`. Shouldn't it go to `TensorFunction`?

Reply:
This is a good point. This will require some more in depth thinking, as it will depend whether or not the Function is constructed from a Grid or from dimensions.

Reply:
Do you suggest to hold for that? 
What's wrong in moving this into the superclass, even though it's not super bullet proof? it'd be at least better than "hiding" it in the subclass...

Reply:
In the super class it would end up with some weird if/else as Function need `self.grid.indices` as aliases while `Sparse` needs `self.indice` which would need to check the type of input.

Reply:
just add a new private property in the superclass and override it in the subclass. Please, use the following format for the name: ``_arg_<name>``, for example ``_arg_aliases``, because that's what I'm going to use anyway as of #507